### ISSUE TITLE
SSH2->read performance question

### ISSUE THREAD

**Issue by metaclassing** [phpseclib/phpseclib#670]
Good afternoon, I have a device (Cisco ASA firewall) giving me quite a bit of grief. I found that commands with short output (60kb) were returning very quickly, but longer output (600kb) took quite a bit longer than expected to come back.

I added some debugging to record performance of the function and loaded the numbers into Excel to graph. I'm trying to attach a picture here:
![netsshread](https://cloud.githubusercontent.com/assets/1027733/7420881/ef8d1854-ef45-11e4-95ee-3fc8b86e9ad7.png)

The output returned from the device is a sequence of ~3400 lines, each 175 bytes in length.

The Net/SSH2.php read function performs a while loop reading data from _get_channel_packet and adding the results to $this->interactiveBuffer.

The first thousand iterations through this loop take about 4 seconds. If the performance were linear I would expect a total run time of <15 seconds. What I am seeing are run times > 3 times that expected value, as with each iteration through the loop things get slower and slower.

Because every line of text returned from the device is ~ the same length (174-175 bytes per line), the amount of data moved in every iteration is very comprable. The only difference seems to be the amount of time taken to process the same data increases.

I have tested OTHER devices (Cisco IOS-XE routers, Nexus NXOS switches) and seen many magnitudes of performance better than these firewalls. The issue appears to be platform specific, and possibly related to how data is being sent from the ASA to the Net_SSH2 library.

Are there any settings or tweaks regarding the SSH connection or channel itself, or changes I could make to the library to wait for a larger buffer of input to return instead of 175 bytes at a time, or anything else you could think of to possibly alleviate my challenge?

I really love this library and use it extensively, thanks for your time and help.

**Comment by terrafrost** [comment]
It's not entirely clear to me how to interpret that graph. The y axis represents the number of read iterations. At least on the right hand side it does. What do the numbers on the left hand side signify? And what does the x access represent?

Anyway, `read()` makes calls `_get_channel_packet()` in a while loop. The size of this packet can be influenced by the window size (server to client), which can't really be controlled by the client, and the packet size, which by default, is 0x4000 or 16,384.

My guess is that the problem is with the window size. I've seen SSH servers start off with a big one and just gradually throttle it down to a very small number. Older versions of phpseclib had issues with calculating the window size but 0.3.10 (or the latest git version) should remedy those issues.

Anyway, to make any definitive diagnosis, logs would help. Maybe just 10mb of logs idk. If you could get those and send them to me, somehow, that'd be helpful. If they're small enough you can put them in a pastebin. If they're too big you can email them to terrafrost@php.net.

Thanks!

**Comment by metaclassing** [comment]
Sorry for the lack of clarity regarding the picture. There are two data sets and two axis. On the left is the axis for the blue graph, bytes per second, on the right is the axis for the orange graph, iterations of the while loop per second. The X axis is time, each cycle was counted with bytes accumulated with microtime stamping.

I will do a packet capture this afternoon and see if I can determine window size of the TCP segments being used to gather the information. My fear is that the SSH server on the ASA is doing something profoundly dumb like sending each line in its own packet, but will capture to confirm.

What should I define/set to generate the best logging output for you? Thanks!

**Comment by metaclassing** [comment]
Well, here is the wireshark image. Let me know if you want the actual PCAP file itself...
![wireshark](https://cloud.githubusercontent.com/assets/1027733/7435749/fed56c3c-f00b-11e4-8e8c-4e174a6f5525.png)
It looks like the window is ok (32k) and each packet within the stream is small (I guess this is to be expected with 175 bytes of data per packet, then encrypted)

More interestingly it looks like the ASA finishes sending the data to phpseclib in ~12 seconds of actual transfer time (running command and sending output back). Then in the packet capture there is no traffic for ~28 more seconds while PHP is in that while(1) loop doing the _get_channel_packet() piece.

I flipped on realtime debug to a file (defined to 4) and will email you the file. Thanks!

**Comment by terrafrost** [comment]
Well I'm not seeing any packets that'd change the window size in the log that I got. It's not showing the initial channel opening packets either so idk what the initial packet sizes or window sizes are. I guess I could modify phpseclib to log specifically that instead of all the raw data but...

> More interestingly it looks like the ASA finishes sending the data to phpseclib in ~12 seconds of actual transfer time (running command and sending output back). Then in the packet capture there is no traffic for ~28 more seconds while PHP is in that while(1) loop doing the _get_channel_packet() piece.

Presumably if there's no traffic for ~28 seconds the script isn't stuck in the while loop - presumably it's blocking. Maybe on the `fread()` in `_get_binary_packet()`, which in turn, is called by `_get_channel_packet()`, or maybe the `stream_select()` in `_get_channel_packet()`. But blocking either way.

The way this paragraph is describing the problem makes me think that the problem may be that what `read()` is expecting is wrong. Maybe the case is wrong or maybe you think you're passing in a regular expression when you're not. Can you post the code that you're using?

Thanks!

**Comment by metaclassing** [comment]
OK, I have written a simple test script and grabbed the latest code off GitHub without any of my abstraction layer in the middle. I can reliably duplicate the behavior I described in my graph and packet capture.

My test script is on pastebin here: http://pastebin.com/KaSL7WTb

What I am seeing from a time/output perspective is as follows:
# php test.php
Got 79 bytes in 0.09109902381897 seconds.
Got 56 bytes in 0.12536311149597 seconds.
Got 45 bytes in 0.12801694869995 seconds.
Got 1718 bytes in 0.1345579624176 seconds.
Got 53141 bytes in 0.54392099380493 seconds.
Got 496745 bytes in 29.002861022949 seconds.

Running a live tcpdump on the machine executing test.php I see that packets from the SSH server (ASA firewall) complete returning the output of the command much faster (12-14 seconds) than phpseclib's SSH2->read() function is picking them up from the buffer.

**Comment by metaclassing** [comment]
I modified the latest github code to add a performance counter to the ssh2->read function while(1) loop: Please take a look at my results here: http://pastebin.com/Bbpmnx4F
I am wondering what the operation in use is that slows by a factor of ~100 as the interactiveBuffer size grows.

**Comment by terrafrost** [comment]
That makes sense. Can I ask what you're using as the regular expression? If it's something like a.*b then there could in theory be 500,000 bytes between a and b. But maybe I could just add a disclaimer to the comment saying that the regular expression is only checked against the last 1,000 bytes or so.

**Comment by metaclassing** [comment]
The short answer to your first question is: Network & telecom equipment vendors are very inconsistent in their prompt format, and many devices have polymorphic prompts that change continually throughout an interactive shell session with the device.

An image helps explain the predicament we frequently run into:
![prompt](https://cloud.githubusercontent.com/assets/1027733/7495987/05242442-f3d7-11e4-9490-b4fd7af0eba2.png)

A little background/history: Originally many of my management & provisioning tools were written in the 2008 time frame in Perl with Net_SSH, with a bolt-on PHP web interface that ran the Perl scripts.

In ~2010 they were re-written entirely in PHP using the PHP SSH PECL extension which worked, but had some limitations regarding flexible prompt matching of devices.

Eventually, I ran across your excellent library and we ripped out all of the non-native PHP functionality, which made all of the software cross platform compatible (ran well on both Linux and Windows) and its use grew.

Today I and others use it to automate and manage thousands of network devices across many various vendors. The complexity of different equipment lead to the creation of a wrapper library to handle prompt detection (assuming the device has never been logged into before) and some creative workarounds.

The primary challenge is that each device vendor, type, and OS version potentially behave differently. I have different REGEX for different device and OS types with logic for vendor & OS detection on the fly:
```
        $this->patterns = array(
            /*
                Sample Prompts: ( Test with http://regex101.com/ )
                    IOS  -   KHONEMDCRRR01#
                    IOS-XE  -   KHONEMDCRWA02#
                    IOS-XR  -   RP/0/RSP0/CPU0:KHONEMDCRWA01#
                    NXOS    -   KHONEMDCSWC01_ADMIN#
                    ASA  -   khonedmzrfw01/pri/act/901-IN#
            */
                array(  'devicetype'    => 'ciscoxr' ,
                    'detect'    => '/RP\/0\/RSP0\/CPU0:([\w\-]+)(.*)[#>]\s*$/' ,
                    'match'     => '/(.*)RP\/0\/RSP0\/CPU0:%s.*(>|#)\s*/'
                    ),

                array(  'devicetype'    => 'cisco' ,
//                  'detect'    => '/([\w\-]+)(\/.*)?[#>]\s*$/' ,
                    'detect'    => '/(?!.*:)([\w\-\/]+)[#>]\s*$/' ,
                    /*
                                    ^--- Dont match anything up to a leading : (XR format)
                                            ^--- Match a-z0-9 - and /
                                                        ^--- Terminate match with our prompt enders > and #
                                                            ^--- Ignore any trailing whitespace
                    */
                    'match'     => '/(.*)%s.*(>|#)\s*/'
                    ),

              array(  'devicetype'    => 'cisconxos' ,
                    'detect'    => '/([\w\-]+)(\/.*)?[#>]\s*$/' ,
                    'match'     => '/(.*)%s.*(>|#)([^ \n\r^M]+)/'
                    )
```

In regard to a more or less official performance recommendation, the biggest benefit might come from changing a single line:
```
preg_match($expect, $this->interactiveBuffer, $matches);
```
to
```
preg_match($expect, substr($this->interactiveBuffer, -1024), $matches);
```
I have not thoroughly tested this change to the library's logic, and someone more familiar with the internal workings of PHP itself would have to weigh in on the function or logic resulting in the best performance.

My fear in suggesting such a change is unintended consequences for other users with vastly different use cases. A disclaimer may be an effective mechanism listing the assumption that a prompt match regex will never be >1024 characters long, however with my limited knowledge of the greater community of users I wouldn't presume to know best.

Thanks again for all the work maintaining this code base, it has saved many thousands of man-hours of effort making repetitive changes and mass desired-state configuration management and auditing. Please feel free to close this issue on Github, I made changes locally to my copy of SSH2.php and am pleased with the performance improvements.

**Comment by bantu** [comment]
This should be fixed by #737
